File: api/includes/body_composition_calculator.php
<?php
// Body Composition / Body Fat calculation core — single source of truth.
// Used by save.php, update.php and bulk-save.php so every code path produces
// identical results. Client-side (Dart) mirrors this only for a live preview;
// whatever this file returns is what actually gets persisted.

/** Attempts with a spread over this many mm get a non-blocking "recheck" flag. */
const BC_ATTEMPT_SPREAD_LIMIT_MM = 2.0;
const BC_CALC_COMPLETE = 'COMPLETE';
const BC_CALC_INCOMPLETE_SKINFOLD = 'INCOMPLETE_SKINFOLD';
const BC_DEFAULT_REQUIRED_SITES = ['biceps', 'triceps', 'subscapular', 'suprailiac'];

/**
 * Age in whole years at the assessment date (not "today") — a player's age
 * band must reflect how old they were when measured, since historical
 * assessments are never recomputed when formulas change later.
 */
function bc_calc_age(string $dob, string $assessmentDate): ?int {
    try {
        $birth = new DateTime($dob);
        $at    = new DateTime($assessmentDate);
    } catch (Exception $e) {
        return null;
    }
    if ($at < $birth) return null;
    return $birth->diff($at)->y;
}

/**
 * Averages up to 3 skinfold attempts for one site. Returns null if none of
 * the attempts were provided (site not measured) — callers must not treat
 * that the same as a genuine zero.
 */
function bc_average_attempts(?float $a1, ?float $a2, ?float $a3): ?float {
    $vals = array_values(array_filter([$a1, $a2, $a3], fn($v) => $v !== null));
    if (!$vals) return null;
    return round(array_sum($vals) / count($vals), 1);
}

function bc_attempt_valid(?float $value): bool {
    return $value === null || ($value >= 2.0 && $value <= 60.0);
}

/**
 * True if any two provided attempts for a site differ by more than the
 * acceptable tolerance — surfaced to the caller as a non-blocking warning,
 * never blocks saving (a coach may only be able to take one attempt).
 */
function bc_attempt_spread_flag(?float $a1, ?float $a2, ?float $a3): bool {
    $vals = array_values(array_filter([$a1, $a2, $a3], fn($v) => $v !== null));
    if (count($vals) < 2) return false;
    return (max($vals) - min($vals)) > BC_ATTEMPT_SPREAD_LIMIT_MM;
}

/**
 * Selects the active formula row for the given age/gender. Returns null when
 * no age band covers this age (e.g. under 17 or over 39) — caller must not
 * calculate body fat % in that case, only store the raw measurements.
 */
function bc_select_formula(PDO $pdo, int $age, string $gender = 'male'): ?array {
    $stmt = $pdo->prepare(
        'SELECT * FROM body_composition_formulas
         WHERE gender = ? AND min_age <= ? AND max_age >= ? AND is_active = 1
         ORDER BY version DESC LIMIT 1'
    );
    $stmt->execute([$gender, $age, $age]);
    $row = $stmt->fetch(PDO::FETCH_ASSOC);
    return $row ?: null;
}

function bc_formula_required_sites(?array $formula): array {
    if (!$formula || empty($formula['required_sites'])) return BC_DEFAULT_REQUIRED_SITES;
    $sites = array_values(array_filter(array_map(
        fn($site) => trim((string)$site),
        explode(',', (string)$formula['required_sites'])
    )));
    return $sites ?: BC_DEFAULT_REQUIRED_SITES;
}

/**
 * @param array<string,?float> $averages
 */
function bc_skinfold_completeness(array $averages, ?array $formula): array {
    $required = bc_formula_required_sites($formula);
    $missing = [];
    foreach ($required as $site) {
        if (!array_key_exists($site, $averages) || $averages[$site] === null) {
            $missing[] = $site;
        }
    }
    return [
        'status' => $missing ? BC_CALC_INCOMPLETE_SKINFOLD : BC_CALC_COMPLETE,
        'required_sites' => $required,
        'missing_sites' => $missing,
        'measurement_completeness' => $required
            ? round((count($required) - count($missing)) / count($required), 4)
            : 0.0,
    ];
}

/**
 * BodyFat% = coefficient_a * LOG10(skinfold_sum) + coefficient_b.
 * Must use log10() — NOT log() (natural log) — to match the reference
 * Durnin & Womersley tables. Returns ['error' => ...] when no formula
 * matches this age/gender combination (age band not supported yet).
 */
function bc_calculate_body_fat(PDO $pdo, float $skinfoldSum, int $age, string $gender = 'male'): array {
    if ($skinfoldSum <= 0) {
        return ['error' => 'Skinfold sum must be positive'];
    }
    $formula = bc_select_formula($pdo, $age, $gender);
    if (!$formula) {
        return ['error' => "No body-fat formula supports age $age ($gender)"];
    }
    $bodyFatPercentage = round(
        ((float)$formula['coefficient_a']) * log10($skinfoldSum) + ((float)$formula['coefficient_b']),
        2
    );
    return [
        'body_fat_percentage' => $bodyFatPercentage,
        'formula_code'        => $formula['formula_code'],
        'age_group'           => $formula['formula_name'],
        'formula_version'     => (string)($formula['version'] ?? '1'),
    ];
}

function bc_bmi(float $weightKg, float $heightCm): float {
    $heightM = $heightCm / 100;
    return round($weightKg / ($heightM * $heightM), 2);
}

function bc_fat_mass_kg(float $weightKg, float $bodyFatPercentage): float {
    return round($weightKg * $bodyFatPercentage / 100, 2);
}

function bc_fat_free_mass_kg(float $weightKg, float $fatMassKg): float {
    return round($weightKg - $fatMassKg, 2);
}

File: lib/utils/body_composition_calculator.dart
import 'dart:math' as math;

/// Live-preview mirror of api/includes/body_composition_calculator.php.
/// Only ever used for instant UI feedback before saving — the backend's
/// computed result from save.php/update.php is always what gets persisted
/// and displayed afterward, so this and the PHP version drifting slightly
/// is not a correctness risk, only a preview-accuracy one.
class BodyCompositionFormula {
  final String code;
  final String name;
  final int minAge;
  final int maxAge;
  final double coefficientA;
  final double coefficientB;

  const BodyCompositionFormula({
    required this.code,
    required this.name,
    required this.minAge,
    required this.maxAge,
    required this.coefficientA,
    required this.coefficientB,
  });
}

class BodyCompositionCalculator {
  static const double attemptSpreadLimitMm = 2.0;

  static const List<BodyCompositionFormula> _formulas = [
    BodyCompositionFormula(code: 'age_17_19', name: 'Durnin-Womersley 17-19', minAge: 17, maxAge: 19, coefficientA: 27.409, coefficientB: -26.789),
    BodyCompositionFormula(code: 'age_20_29', name: 'Durnin-Womersley 20-29', minAge: 20, maxAge: 29, coefficientA: 27.775, coefficientB: -27.203),
    BodyCompositionFormula(code: 'age_30_39', name: 'Durnin-Womersley 30-39', minAge: 30, maxAge: 39, coefficientA: 26.781, coefficientB: -27.203),
  ];

  static double log10(double x) => math.log(x) / math.ln10;

  static int? calcAge(DateTime dob, DateTime assessmentDate) {
    if (assessmentDate.isBefore(dob)) return null;
    int age = assessmentDate.year - dob.year;
    final hadBirthdayThisYear = (assessmentDate.month > dob.month) ||
        (assessmentDate.month == dob.month && assessmentDate.day >= dob.day);
    if (!hadBirthdayThisYear) age--;
    return age;
  }

  static double? averageAttempts(List<double?> attempts) {
    final vals = attempts.whereType<double>().toList();
    if (vals.isEmpty) return null;
    return double.parse((vals.reduce((a, b) => a + b) / vals.length).toStringAsFixed(1));
  }

  static bool attemptSpreadFlag(List<double?> attempts) {
    final vals = attempts.whereType<double>().toList();
    if (vals.length < 2) return false;
    final maxV = vals.reduce((a, b) => a > b ? a : b);
    final minV = vals.reduce((a, b) => a < b ? a : b);
    return (maxV - minV) > attemptSpreadLimitMm;
  }

  static BodyCompositionFormula? selectFormula(int age) {
    for (final f in _formulas) {
      if (age >= f.minAge && age <= f.maxAge) return f;
    }
    return null;
  }

  /// Returns null if no formula covers this age — caller should show
  /// "unsupported age band" rather than a fabricated number.
  static double? calculateBodyFatPercentage(double skinfoldSum, int age) {
    if (skinfoldSum <= 0) return null;
    final formula = selectFormula(age);
    if (formula == null) return null;
    final bf = formula.coefficientA * log10(skinfoldSum) + formula.coefficientB;
    return double.parse(bf.toStringAsFixed(2));
  }

  static double bmi(double weightKg, double heightCm) {
    final heightM = heightCm / 100;
    return double.parse((weightKg / (heightM * heightM)).toStringAsFixed(2));
  }

  static double fatMassKg(double weightKg, double bodyFatPercentage) {
    return double.parse((weightKg * bodyFatPercentage / 100).toStringAsFixed(2));
  }

  static double fatFreeMassKg(double weightKg, double fatMassKg) {
    return double.parse((weightKg - fatMassKg).toStringAsFixed(2));
  }
}
